### PR TITLE
Models section

### PR DESCRIPTION
**Pull request by JonnySpruce**

### CODE REVIEW

File: blog/sudoko_blog_structure.md [hunk 1]
```diff
@@ -103,42 +107,79 @@ training_data, validation_data = random_split(dataset, [0.8, 0.2], generator=gen
...
+
+1. On average, how many numbers the model accurately guessed vs. the solution
+2. What percentage of puzzles the model can solve 100% correctly (no incorrect numbers whatsoever)
+
+We planned to train our models against our standard loss calculations, and then once trained, verify the performance using these two metrics. We would run the trained model against the validation dataset to ensure it had never seen any of the puzzles before.
+
+We also read about an interesting technique to improve model performance in sudoku from a GitHub repository called [Can Convolutional Neural Networks Crack Sudoku Puzzles?](https://github.com/Kyubyong/sudoku), where they only take the single number the model is most sure about, update the input with this number and then feed this updated model back into the model, until the puzzle is completely filled out. We decided to also measure the model performance when used this way as well, which we will refer to as "iterative guesses" in the performance tables throughout this blog.
+
+So how did our `MLP` model do? Well, after being trained on over 6 million sudoku puzzles, we put it to the test and you can see the results in the table below:
```
Comment (jstrong-scottlogic) [inline blog/sudoko_blog_structure.md]: Does 6 million here refer to merging datasets?

File: blog/sudoko_blog_structure.md [hunk 2]
```diff
@@ -154,37 +195,162 @@ You can see that a learning rate that is high trains quickly and then struggles
...
+
 ## Batch Normalisation
 
 What is normalisation - what is batch version? Reduces effect of internal covariate shift in model parameters which is changes in their distributions. Makes training quicker and less all over the place (technical term).
 
+![Alt text](0738131a-92ee-4b18-ba15-206ef00c78c4.png)
+
 ## Optimiser
 
 The optimiser is involved in updating the parameters of the model during the training process. It aims to minimise the loss function...
```
Comment (jstrong-scottlogic) [inline blog/sudoko_blog_structure.md]: This does not seem up to date with main.

File: blog/sudoko_blog_structure.md [hunk 3]
```diff
@@ -154,37 +195,162 @@ You can see that a learning rate that is high trains quickly and then struggles
...
+
+    def forward(self, x):
+        mod_x = x.permute([0,2,1])
+        mod_x = mod_x.unflatten(2,(9,9))
+        logits = self.conv_relu_stack(mod_x)
+        logits = logits.squeeze().unflatten(1, (81,9))
+        return logits
+```
+
+We added an additional `Conv2d` layer at the end to make the output have a shape of 1x1, with 729 nodes. This was done because the expected output of the model is 729 numbers (9x9 grid with 1-hot encoding for numbers 1-9 = 9x9x9 = 729).
```
Comment (jstrong-scottlogic) [inline blog/sudoko_blog_structure.md]: It seems we use one-hot in the rest of the blog, should keep consistent.

File: blog/sudoko_blog_structure.md [hunk 4]
```diff
@@ -154,37 +195,162 @@ You can see that a learning rate that is high trains quickly and then struggles
...
+#### Encoding the rules
+
+It appeared that encoding the rules into our model helped it learn more effectively. We decided to take this one step further and encode the rules for the rows and columns as well. We did this by creating two additional convolution layers which were also passed the full puzzle input. The first kernel size was 1x9 to simulate the columns, and the second was 9x1 to simulate the rows. These kernel sizes are illustrated below:
+
+<div align="center">
+    <img src="./images/oneNine.gif" height=200px />
+    <img src="./images/nineOne.gif" height=200px />
+</div>
+
+ Since we wanted the model to be able to take the inputs from all three kernels and then make decisions on what the output should be, we concatenated the outputs of the three layers into a big list of numbers and then passed this into a hidden linear layer. We then had an output layer that had the expected number of nodes for the 1-hot encoded output.
```
Comment (jstrong-scottlogic) [inline blog/sudoko_blog_structure.md]: 1-hot once more.

File: blog/sudoko_blog_structure.md (hunk 1 above)
Comment (JonnySpruce) [reply blog/sudoko_blog_structure.md]: No, it's 2 epochs of 4m (since 20% is the validation set)

Comment (JonnySpruce) [reply blog/sudoko_blog_structure.md]: I'll make that clearer

**Review by jstrong-scottlogic — APPROVED** [review]
